File: python/datastructure/otterop/datastructure/list.py
from otterop.lang.array import Array as _Array
from otterop.lang.generic import Generic as _Generic
from otterop.lang.panic import Panic as _Panic
from otterop.lang.pure_iterator import PureIterator as _PureIterator
from otterop.lang.string import String as _String
from otterop.datastructure._list_iterator import ListIterator as _ListIterator

class List:
                
    def __init__(self):
        self._size = 0
        self._capacity = 4
        generic_t = _Generic()
        self._t_zero = generic_t.zero()
        self._array = _Array.new_array(self._capacity, self._t_zero)

    def _ensure_capacity(self, capacity):
        if self._capacity < capacity:
            self._capacity = self._capacity * 2
            new_array = _Array.new_array(self._capacity, self._t_zero)
            _Array.copy(self._array, 0, new_array, 0, self._size)
            self._array = new_array

    def add(self, element):
        self._ensure_capacity(self._size + 1)
        self._array.set(self._size, element)
        self._size += 1

    def add_array(self, src):
        self._ensure_capacity(self._size + src.size())
        _Array.copy(src, 0, self._array, self._size, src.size())
        self._size += src.size()

    def add_list(self, src):
        self.add_array(src._array)

    def _check_index_out_of_bounds(self, index):
        if index < 0 or index > self._size:
            _Panic.index_out_of_bounds(_String.wrap("index is outside list bounds"))

    def insert(self, index, element):
        self._check_index_out_of_bounds(index)
        self._ensure_capacity(self._size + 1)
        if index < self._size:
            _Array.copy(self._array, index, self._array, index + 1, self._size - index)
        self._array.set(index, element)
        self._size += 1

    def insert_array(self, index, src):
        self._check_index_out_of_bounds(index)
        self._ensure_capacity(self._size + src.size())
        if index < self._size:
            _Array.copy(self._array, index, self._array, index + src.size(), self._size - index)
        _Array.copy(src, 0, self._array, index, src.size())
        self._size += src.size()

    def insert_list(self, index, src):
        self.insert_array(index, src._array)

    def get(self, index):
        self._check_index_out_of_bounds(index)
        return self._array.get(index)

    def remove_index(self, index):
        self._check_index_out_of_bounds(index)
        ret = self._array.get(index)
        if index + 1 < self._size:
            _Array.copy(self._array, index + 1, self._array, index, self._size - index - 1)
        self._size -= 1
        return ret

    def remove_range(self, index, count):
        self._check_index_out_of_bounds(index)
        if index + count > self._size:
            count = self._size - index
        ret = List()
        removed = _Array.new_array(count, self._t_zero)
        _Array.copy(self._array, index, removed, 0, count)
        ret.add_array(removed)
        if index + count < self._size:
            _Array.copy(self._array, index + count, self._array, index, self._size - index - count)
        self._size = self._size - count
        return ret

    def size(self):
        return self._size

    def oop_iterator(self):
        return _ListIterator(self)

    def __iter__(self):
        return _PureIterator.new_iterator(self.oop_iterator())

File: ts/datastructure/otterop/datastructure/List.ts
import { Array } from '@otterop/lang/Array';
import { Generic } from '@otterop/lang/Generic';
import { Panic } from '@otterop/lang/Panic';
import { String } from '@otterop/lang/String';

export class List<T> {

    #array : Array<T>;

    #capacity : number;

    #size : number;

    #tZero : T;

    public constructor() {
        this.#size = 0;
        this.#capacity = 4;
        let genericT : Generic<T> = new Generic<T>();
        this.#tZero = genericT.zero();
        this.#array = Array.newArray(this.#capacity, this.#tZero);
    }

    #ensureCapacity(capacity : number) : void {
        
        if (this.#capacity < capacity) {
            this.#capacity = this.#capacity * 2;
            let newArray : Array<T> = Array.newArray(this.#capacity, this.#tZero);
            Array.copy(this.#array, 0, newArray, 0, this.#size);
            this.#array = newArray;
        }
    }

    public add(element : T) : void {
        this.#ensureCapacity(this.#size + 1);
        this.#array.set(this.#size, element);
        this.#size++;
    }

    public addArray(src : Array<T>) : void {
        this.#ensureCapacity(this.#size + src.size());
        Array.copy(src, 0, this.#array, this.#size, src.size());
        this.#size += src.size();
    }

    public addList(src : List<T>) : void {
        this.addArray(src.#array);
    }

    #checkIndexOutOfBounds(index : number) : void {
        
        if (index < 0 || index > this.#size) {
            Panic.indexOutOfBounds(String.wrap("index is outside list bounds"));
        }
    }

    public insert(index : number, element : T) : void {
        this.#checkIndexOutOfBounds(index);
        this.#ensureCapacity(this.#size + 1);
        
        if (index < this.#size) {
            Array.copy(this.#array, index, this.#array, index + 1, this.#size - index);
        }
        this.#array.set(index, element);
        this.#size++;
    }

    public insertArray(index : number, src : Array<T>) : void {
        this.#checkIndexOutOfBounds(index);
        this.#ensureCapacity(this.#size + src.size());
        
        if (index < this.#size) {
            Array.copy(this.#array, index, this.#array, index + src.size(), this.#size - index);
        }
        Array.copy(src, 0, this.#array, index, src.size());
        this.#size += src.size();
    }

    public insertList(index : number, src : List<T>) : void {
        this.insertArray(index, src.#array);
    }

    public get(index : number) : T {
        this.#checkIndexOutOfBounds(index);
        return this.#array.get(index);
    }

    public removeIndex(index : number) : T {
        this.#checkIndexOutOfBounds(index);
        let ret : T = this.#array.get(index);
        
        if (index + 1 < this.#size) {
            Array.copy(this.#array, index + 1, this.#array, index, this.#size - index - 1);
        }
        this.#size--;
        return ret;
    }

    public removeRange(index : number, count : number) : List<T> {
        this.#checkIndexOutOfBounds(index);
        
        if (index + count > this.#size) {
            count = this.#size - index;
        }
        let ret : List<T> = new List<T>();
        let removed : Array<T> = Array.newArray(count, this.#tZero);
        Array.copy(this.#array, index, removed, 0, count);
        ret.addArray(removed);
        
        if (index + count < this.#size) {
            Array.copy(this.#array, index + count, this.#array, index, this.#size - index - count);
        }
        this.#size = this.#size - count;
        return ret;
    }

    public size() : number {
        return this.#size;
    }
}

